src/main/java/org/candlepin/auth/ActivationKeyAuth.java

```diff
@@ -0,0 +1,163 @@
+ * in this software or its documentation.
+ */
+package org.candlepin.auth;
+
+import org.candlepin.exceptions.NotAuthorizedException;
+import org.candlepin.model.OwnerCurator;
+import org.candlepin.model.activationkeys.ActivationKey;
+import org.candlepin.model.activationkeys.ActivationKeyCurator;
+
+import com.google.inject.Inject;
+
+import org.jboss.resteasy.spi.HttpRequest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xnap.commons.i18n.I18n;
+
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.inject.Provider;
+import javax.ws.rs.core.MultivaluedMap;
+
+
+/**
+ * Auth implementation to allow authentication using activation keys for
+ * the consumer registration.
+ */
+public class ActivationKeyAuth implements AuthProvider {
+
+    private static final Logger log = LoggerFactory.getLogger(ActivationKeyAuth.class);
+
+    private final Provider<I18n> i18nProvider;
+    private final ActivationKeyCurator activationKeyCurator;
+    private final OwnerCurator ownerCurator;
+
+    @Inject
+    ActivationKeyAuth(Provider<I18n> i18nProvider, ActivationKeyCurator activationKeyCurator,
+        OwnerCurator ownerCurator) {
+        this.i18nProvider = Objects.requireNonNull(i18nProvider);
+        this.activationKeyCurator = Objects.requireNonNull(activationKeyCurator);
+        this.ownerCurator = Objects.requireNonNull(ownerCurator);
+    }
+
+    @Override
+    public Principal getPrincipal(HttpRequest httpRequest) {
+        if (!isAllowedEndpoint(httpRequest)) {
+            String msg = this.i18nProvider.get()
+                .tr("Activation key authorization is not available for the requested resource.");
+            log.warn(msg);
+            return null;
+        }
+
+        MultivaluedMap<String, String> queryParams = httpRequest.getUri().getQueryParameters();
+        if (!validateRequiredParams(queryParams)) {
+            return null;
+        }
+
+        String ownerKey = queryParams.getFirst("owner");
+        if (!this.ownerCurator.existsByKey(ownerKey)) {
+            throw notAuthorized("Specified owner does not exist.");
+        }
+
+        String keys = queryParams.getFirst("activation_keys");
+        List<ActivationKey> activationKeys = findActivationKeys(ownerKey, keys);
+        if (activationKeys.isEmpty()) {
+            throw notAuthorized("None of the specified Activation keys were found.");
+        }
+
+        return new ActivationKeyPrincipal();
+    }
+
+    /**
+     * Checks whether the request is made against one of the endpoints for which
+     * we allow activation key authentication.
+     *
+     * @param httpRequest request to check
+     * @return true if activation key auth is allowed for the requested endpoint
+     */
+    private boolean isAllowedEndpoint(HttpRequest httpRequest) {
```

Comment:
Alright. I'll noodle on this for a bit and see if we can find a reasonable middle ground.

Reply:
I added this as a param to the SecurityHole. It's more in line with other params we have there. Also we don't have to check for this in VerifyAuthorizationFilter. ActivationKeyPrincipal has no permissions so it will fail to authorize.